### DISCUSSION TITLE
Tab key behaviour

### DISCUSSION THREAD

**Discussion by kj** [helix-editor/helix#8258] [Q&A] [answered]
The one thing I still really can't get my head around in Helix (and I've been using it for a few years now) is how the tab key behaves, it feels very unintuitive/inconsistent.

* **Tab** on blank line (or before first non-whitespace character)
    * Indent (insert spaces or tab depending on language)
* **Shift+Tab** on blank line (or before first non-whitespace character)
    * Indent (insert spaces or tab depending on language)
* **Tab** after first non-whitespace character
    * No idea, it sometimes does nothing (e.g. in a scratch buffer with no language), or the cursor skips somewhere ahead, sometimes to the end of the line or the next paragraph (something to do with Tree Sitter perhaps?)
* **Shift+Tab** after first non-whitespace character
    * Indent from cursor (insert spaces or tab depending on language)

And then of course it's also used to select a completion if the completion prompt comes up (I'd rather just use Ctrl+n and Ctrl+p myself). What's up with the behaviour in the third point? And how can I just insert a literal tab character?

**Comment by the-mikedavis** [comment, marked as answer]
Assuming you're running on master, that third point is https://github.com/helix-editor/helix/pull/4443. You can turn off smart tab: https://docs.helix-editor.com/master/configuration.html#editorsmart-tab-section

**Reply by kj** [reply]
Ahh, that's it! Thank you (and you're right, I actually forgot I was on master so that would explain why I couldn't find anything in the documentation).

Do you know if it's possible to input a literal tab character in a language that is configured for space indent?

Anyway, this was definitely my main issue. It feels more consistent now with that option disabled.

**Reply by pascalkuthe** [reply]
Inserting tab characters with a different ident setting is not possible but that seems super nieche. You can change the identity setting at any time using `:ident` (or in your config).

**Reply by kj** [reply]
Yeah it probably is actually (I think it felt like more of an issue earlier as I didn't have a clear idea about what was happening).

**Reply by Omnikar** [reply]
If you really need a tab character you can do `!echo -n "\t"` as a workaround.
